fhir-persistence/src/main/java/com/ibm/fhir/persistence/ResourceResult.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * (C) Copyright IBM Corp. 2021, 2022
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.persistence;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.ibm.fhir.model.annotation.Required;
+import com.ibm.fhir.model.resource.Resource;
+
+/**
+ * The base wrapper result wrapper used to represent a resource being returned from a
+ * persistence interaction.
+ * Instances are immutable and can be constructed via {@code new ResourceResult.Builder<T>()}.
+ */
+public class ResourceResult<T extends Resource> {
+    @Required
+    private final T resource;
+    private final boolean deleted;
+    private final String resourceTypeName;
+    private final String logicalId;
+    private final int version;
+    private final Instant lastUpdated;
+
+    /**
+     * Private constructor used by the Builder to create a new ResourceResult instance
+     * @param builder
+     */
+    private ResourceResult(Builder<T> builder) {
+        resource = builder.resource;
+        deleted = builder.deleted;
+        resourceTypeName = builder.resourceTypeName;
+        logicalId = builder.logicalId;
+        version = builder.version;
+        lastUpdated = builder.lastUpdated;
+    }
+
+    /**
+     * Create a Builder for building instances of this class
+     * @param <T>
+     * @return
+     */
+    public static <T extends Resource> ResourceResult.Builder<T> builder() {
+        return new ResourceResult.Builder<>();
+    }
+
+    /**
+     * Convenience method to convert a Resource value to a ResourceResult,
+     * which is useful for unit-tests
+     *
+     * @param <T>
+     * @param resource with a valid meta field
+     * @return
+     */
+    public static ResourceResult<? extends Resource> from(Resource resource) {
+        return ResourceResult.builder()
+                .resource(resource)
+                .lastUpdated(resource.getMeta().getLastUpdated().getValue().toInstant())
+                .version(Integer.parseInt(resource.getMeta().getVersionId().getValue()))
+                .logicalId(resource.getId())
+                .resourceTypeName(resource.getClass().getSimpleName())
+                .build();
+    }
+
+    /**
+     * Convenience function to convert a list of ResourceResults to a list of Resources
+     * which are not null
+     * @param resourceResults
+     * @return
+     */
+    public static List<? extends Resource> toResourceList(List<ResourceResult<? extends Resource>> resourceResults) {
```

Comment:
random thought:  I wonder if this class could help us implement https://github.com/IBM/FHIR/issues/2432 in the future.
is there anything we'd design differently?